### DISCUSSION TITLE
Calculation precharge-frames of XDCAM files

### DISCUSSION THREAD

**Discussion by donkon** [bbc/bmx#23] [General]
Hi Philipp,

we have an issue with the result of mxf2raw and XDCAM files: the presented precharge-frames (e.g. `<precharge>-2</precharge>`) are different than other tools we use (such as mediainfo).

It's about the analysis of XDCAM material that has been precisely clipped frame by frame. Due to the GOP (Group of Pictures) length of 12 frames, precharge frames are created -  usually fewer than 12 frames. However, the ARD's software regularly generates files with precharge frames more than 12 frames, up to 14 (12 - 14).

In this case, we obtain incorrect results. For example Legalizer delivers accurate precharge-frame results, mediainfo shows accurate material- and source-timecode. The examples in the Folders make the problem clear: The mxf2raw-Analysis delivers too few Precharge-Frames, in both cases "-2" instead of 12 and 14 Frames (obviously the first 12 Frames are neglected).

We have a example here ([the URL pointing to the file](https://storage.googleapis.com/video-content-stp-sz-mfs/media/wdr/analyse/BMX_Examples/14_Precharge-Frames/586364_0739477_CS_278517_-_11-00-00-14_bis_11-00-59-16.mxf?x-goog-signature=03afb363ced146d0865611e743e9e5d8babdecd944cf3233bbb9a5709c3b98111c73d51262fe66de0a9e5d97197684c89209520b80e4e52ef9c80ffa906e14f133dff20ac994415019314854024f6bbba59a79071dbc49e65677624cf4dba5b8ad9c1bb8fded69b89b98fbc2df237e5fb88af4e1855625ae2ce0847184638dd840a57d891db4ec9e669176ccaa2dcada7e5462d14eb2106f6a136e982118a4c879b77ff604f32d1a54797b809c4ae9f6094ff3b8472705ca9d113457466f862e966e157e77b7e7f166aaa2ed43a01a69dc7c41b0a318c9facebd3a88bc312008f8ac023cfd09aa3feb3e82fb6ae203a220dc37c704a55ea79394b607b809b18b&x-goog-algorithm=GOOG4-RSA-SHA256&x-goog-credential=storage-access%40project-ard-stp-sz-mfs-int-gcp.iam.gserviceaccount.com%2F20230803%2Feurope-west3%2Fstorage%2Fgoog4_request&x-goog-date=20230803T072146Z&x-goog-expires=604800&x-goog-signedheaders=host) is valid for 7 days):

`mxf2raw --info-format xml -i "https://storage.googleapis.com/video-content-stp-sz-mfs/media/wdr/analyse/BMX_Examples/14_Precharge-Frames/586364_0739477_CS_278517_-_11-00-00-14_bis_11-00-59-16.mxf?x-goog-signature=03afb363ced146d0865611e743e9e5d8babdecd944cf3233bbb9a5709c3b98111c73d51262fe66de0a9e5d97197684c89209520b80e4e52ef9c80ffa906e14f133dff20ac994415019314854024f6bbba59a79071dbc49e65677624cf4dba5b8ad9c1bb8fded69b89b98fbc2df237e5fb88af4e1855625ae2ce0847184638dd840a57d891db4ec9e669176ccaa2dcada7e5462d14eb2106f6a136e982118a4c879b77ff604f32d1a54797b809c4ae9f6094ff3b8472705ca9d113457466f862e966e157e77b7e7f166aaa2ed43a01a69dc7c41b0a318c9facebd3a88bc312008f8ac023cfd09aa3feb3e82fb6ae203a220dc37c704a55ea79394b607b809b18b&x-goog-algorithm=GOOG4-RSA-SHA256&x-goog-credential=storage-access%40project-ard-stp-sz-mfs-int-gcp.iam.gserviceaccount.com%2F20230803%2Feurope-west3%2Fstorage%2Fgoog4_request&x-goog-date=20230803T072146Z&x-goog-expires=604800&x-goog-signedheaders=host"`

Here are the results from other tools:

- [mediainfo](https://storage.googleapis.com/video-content-stp-sz-mfs/media/wdr/analyse/BMX_Examples/14_Precharge-Frames/MediaInfo_11-00-00-14.json?x-goog-signature=7b698e823ba30f894875420b3a5fe5e590530846152f357060b3c5e86a14befefe40bad3984efa714cc8595a4eb3acc5cbeeff37ed025365f99e1f04dcc2ec9cbf90c492861f2aec326d4dac1df889d377dcd769ef4ec306fe0799af937c713d7c5db2e2ab418d7a29605ff53ddeb7158bdcaa142e0d4c4d74d8013a2e350bbb646b9dbb052e6d8d79e944212dd9f8e1d929a79ff9a44e3de484cf697321c68bee894c4f832068cbda8de3af5d89f10cb9d98f331c395b501da50927be369830580b7305464c0682b30f1df95ff6542c7158f742ef8ee432f777437c89393074869f7fb04dcff6c522facc875c0684e2f5088124458f20c92a6452a9bf68d600&x-goog-algorithm=GOOG4-RSA-SHA256&x-goog-credential=storage-access%40project-ard-stp-sz-mfs-int-gcp.iam.gserviceaccount.com%2F20230803%2Feurope-west3%2Fstorage%2Fgoog4_request&x-goog-date=20230803T073233Z&x-goog-expires=604800&x-goog-signedheaders=host)
- [mxflegalizer](https://storage.googleapis.com/video-content-stp-sz-mfs/media/wdr/analyse/BMX_Examples/14_Precharge-Frames/586364_0739477_CS_278517_-_11-00-00-14_bis_11-00-59-16.mxf.pdf?x-goog-signature=9065096f5a854b57db6d769877f81339ce30bfba17bd108cd44e60b1337d40abe79f7fee1ae4eb3fbeb00bdf830f2998d06592ed5b39a6726d037b8a746acfac28ea6af5d63fc2d27fac350a0ba2b17294032e9f8bf79f3a3409fcc9e6d1154883b747993488c63ba5b3d1f010df318da90197f49723ef23ff18f7ad3caf36b4caddcbee313695a1b576ce906f1be0adb9cdb8a9eb8f182e012d4719b16ed817b90b76a68c2654a2a871a238a4290ebcb90ad39e025edc43e4334cc396eb0449f485d457aade9aeec0c5c77f6a548e4dada66644fd6a24cb3d9ef4db8c05cab7052af4b592d9cb565a6b3cf5fe966e3c6c4e3a40a632c4f6af9803038308b41f&x-goog-algorithm=GOOG4-RSA-SHA256&x-goog-credential=storage-access%40project-ard-stp-sz-mfs-int-gcp.iam.gserviceaccount.com%2F20230803%2Feurope-west3%2Fstorage%2Fgoog4_request&x-goog-date=20230803T073435Z&x-goog-expires=604800&x-goog-signedheaders=host)

**Comment by philipnbbc** [comment]
I've taken a look and I think mxf2raw is reporting correct information for the amount of pre-charge required. mxf2raw reports the (minimum) amount of frames needed to produce the output time range. In this file the output starts at frame 14. The coded frame required to produce output position 14 is frame 12, which is an I-frame. That means there are 2 frames (the pre-charge) needed to decode output frame 14 and subsequent frames.

The legalizer references [0054W - Origin Values](https://qc.ebu.io/items/0054W/) in the EBU QC and the legalizer names the property "Precharge Frames". The `Origin` value in the file source package is indeed 14.

(The timecodes reported by mxf2raw look fine as well. Note that mxf2raw reports the timecode of the first _output_ frame which is 11:00:00:14 for the file source package)

**Reply by dysty13** [reply]
Hi Philip,
Our Problem is, that we need the 14 Frames from somewhere. mxf2raw doesn`t report the source-timecode 11:00:00:00, (like mediainfo does).
In mxf2raw StartTimecode "material" is always the same like "file_source":
    material        : 11:00:00:14
    file_source     : 11:00:00:14
    
And mxf2raw doesn`t report the Origin value "14 Frames". So all we have is 2 Precharge-Frames, that doesn`t match with our file.
Is it possible to add the value "origin" in mxf2raw (14 in our case)? Or report the source`s StartTimecode (11:00:00:00 in our case)?

Thanks!

**Comment by philipnbbc** [comment]
I think adding something like "origin" would generally be useful and I'll look at adding it or something equivalent to tell you how many additional frames there are. Not yet sure whether it should be as explicit about the property name "origin" or whether something like "precharge_available" is better to deal with complexities of different track "Origin" property values (e.g. in OPAtom). A similar thing applies to the file source package timecodes.

This also points to the fact the bmx is more focused on wrapping and trans-wrapping files than on validation or file analysis. mediainfo and other validation / analysis tools could provide more direct information with less interpretation than bmx does.

**Comment by philipnbbc** [comment]
I've created https://github.com/bbc/bmx/pull/26 to add properties that you can use, namely `max_available_precharge`. Could you take a look and let me know if it is ok?

It's maybe not quite as elegant as the output of mxfflegalizer for example but works for how bmx has been implemented. There are invariably some corner cases it doesn't handle but it should work fine for OP1a.

**Reply by dysty13** [reply]
That's great, thank you for your effort. This is exactly what we need. We will integrate it into our Workflow as soon as possible and will let you know if it works,

**Comment by dysty13** [comment]
Hi Philip,
We have tested the new field and as expected, it works very well . It would be great, if you merge it into the mainbranch.

**Reply by philipnbbc** [reply]
Merged - thanks for testing.
